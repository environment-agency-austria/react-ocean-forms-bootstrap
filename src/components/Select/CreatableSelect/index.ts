export * from './CreatableSelect';
export * from './CreatableSelect.types';
