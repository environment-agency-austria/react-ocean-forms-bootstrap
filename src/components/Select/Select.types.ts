import { IFieldComponentProps } from 'react-ocean-forms';
import { IInfoProps } from '../FieldLine';

/**
 * Selectable option
 */
export interface ISelectOption {
  /**
   * Display label
   */
  label: string;
  /**
   * Internal value
   */
  value: string;
}

// tslint:disable-next-line:no-any
export function isSelectOption(object: any): object is ISelectOption {
  return object && typeof object.label === 'string' && typeof object.value === 'string';
}

/**
 * Type for an array of ISelectOption
 * @see ISelectOption
 */
export type ISelectOptions = ISelectOption[];

/**
 * Props for the Select component
 */
export interface ISelectProps extends IFieldComponentProps, IInfoProps {
  /**
   * Placeholder string that will be displayed while
   * the component is empty
   */
  placeholder: string;
  /**
   * True, if the select component should be in multi-select mode
   */
  multi: boolean;
  /**
   * Selectable options
   */
  options: ISelectOptions;
}
