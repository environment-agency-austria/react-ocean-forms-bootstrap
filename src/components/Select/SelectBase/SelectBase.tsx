/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { useCallback, useMemo } from 'react';
import { Input as StrapInput } from 'reactstrap';

import { useField } from 'react-ocean-forms';
import { ActionMeta } from 'react-select/lib/types';
import { FieldLine } from '../../FieldLine';
import {
  IPreparedSelectProps, ISelectBaseProps, ISelectFieldValue,
  ISelectOption, ISelectOptions, isSelectOption,
} from './SelectBase.types';

/**
 * The type that the field can have
 */
type SelectFieldValue = ISelectFieldValue | undefined;

/**
 * Result of an update
 */
interface IUpdatedLabelsResult {
  /**
   * True, if options were updated
   */
  needsUpdate: boolean;
  /**
   * The options
   */
  selectedOptions: SelectFieldValue;
}

/**
 * Base component for displaying bootstrap
 * form groups with an select input and
 * oForm support
 */
export const SelectBase = <TSubmitValue extends unknown = ISelectFieldValue>(props: ISelectBaseProps<TSubmitValue>): JSX.Element => {
  const {
    renderSelect,
    placeholder = 'ojs_select_placeholder',
    options,

    // booleans
    multi = false,
    loading,
    rtl,
    searchable,
    clearable,

    // Other props (field line)
    ...fieldlineProps
  } = props;

  const { fieldProps, metaProps } = useField(props);

  /**
   * Manually handle the onChange event of
   * the select because oForms is expecting
   * default input onChange behaviour.
   */
  const handleChange = useCallback((value: ISelectFieldValue, action?: ActionMeta): void => {
    const { handleChange } = props;

    fieldProps.onChange({
      target: {
        value,
      },
    });

    if (handleChange) {
      handleChange(value, action);
    }
  }, [fieldProps, props]);

  /**
   * Manually handle the onBlur event of
   * the select because oForms is expecting
   * default input onBlur behaviour.
   */
  const handleBlur = useCallback((): void => {
    fieldProps.onBlur();
  }, [fieldProps]);

  /**
   * Update the label of a select option according to the available options.
   * If null is returned, the label is already up to date.
   *
   * @param options The available select options
   * @param value The value that may need update
   */
  const updateSelectOptionLabel = useCallback((options: ISelectOptions, value: ISelectOption): ISelectOption | undefined => {
    let selectableValue: ISelectOption | undefined;
    options.some(item => {
      if (item.value !== value.value) { return false; }

      selectableValue = item;

      return true;
    });

    if (selectableValue === undefined || selectableValue.label === value.label) {
      return undefined;
    }

    return {
      ...value,
      label: selectableValue.label,
    };
  }, []);

  /**
   * Get the selected options with updated labels.
   * This might be needed if the labels were updated because of a language changes
   * @param options The available options
   * @param value The current value
   */
  const getUpdateLabels = useCallback((options: ISelectOptions, value: SelectFieldValue): IUpdatedLabelsResult => {
    let selectedOptions: SelectFieldValue;
    let needsUpdate = false;

    if (isSelectOption(value)) {
      // If only one option is selected
      const updated = updateSelectOptionLabel(options, value);
      needsUpdate = (updated !== undefined);
      selectedOptions = (needsUpdate)
        ? updated
        : value;
    } else if (Array.isArray(value)) {
      // Map the values
      selectedOptions = value.map(v => {
        const updated = updateSelectOptionLabel(options, v);
        // If null was returned, no updated is needed
        if (updated === undefined) { return v; }
        // Otherwise set the flag to true
        needsUpdate = true;

        return updated;
      });
    }

    return {
      needsUpdate,
      selectedOptions,
    };
  }, [updateSelectOptionLabel]);

  /**
   * Get the value of the field with updated labels
   * @param field The field
   * @param options The available options
   */
  const fieldValue = useMemo(() => {
    const fieldValue = (fieldProps.value as ISelectFieldValue | undefined);
    const {
      needsUpdate,
      selectedOptions,
    } = getUpdateLabels(options, fieldValue);
    if (needsUpdate && selectedOptions !== undefined) {
      // Check if the current value has a different label than the value
      // with the same key in the options array. Bugfix to change the
      // selected label when the current language changes.
      handleChange(selectedOptions);
    }

    return selectedOptions;
  }, [fieldProps.value, getUpdateLabels, handleChange, options]);

  // Support for plaintext display
  if (metaProps.plaintext) {
    let displayValue = '';
    if (isSelectOption(fieldValue)) {
      displayValue = fieldValue.label;
    } else if (Array.isArray(fieldValue)) {
      displayValue = fieldValue.map(v => v.label).join(', ');
    }

    return (
      <FieldLine {...props} fieldProps={fieldProps} metaProps={metaProps}>
        <StrapInput
          {...fieldProps}
          value={displayValue}
          plaintext
          onChange={undefined}
        />
      </FieldLine>
    );
  }

  // Generate a css class based on the validity of the select element
  let selectClass = 'react-select-control';
  selectClass = (selectClass + (metaProps.valid ? '' : ' is-invalid')).trim();

  // Format the placeholder
  const placeholderText = metaProps.stringFormatter(placeholder);

  // Prepare the props for the renderSelect method
  const preparedProps: IPreparedSelectProps = {
    id: fieldProps.id,
    inputId: `${fieldProps.id}-input`,
    value: fieldValue,
    isDisabled: fieldProps.disabled,
    isMulti: multi,
    isRtl: rtl,
    isLoading: loading,
    isSearchable: searchable,
    isClearable: clearable,
    onChange: handleChange,
    onBlur: handleBlur,
    options: options,
    placeholder: placeholderText,
    className: selectClass,
  };

  return (
    <FieldLine
      fieldProps={fieldProps}
      metaProps={metaProps}
      {...fieldlineProps}
    >
      {renderSelect(preparedProps, fieldProps)}
    </FieldLine>
  );
};
