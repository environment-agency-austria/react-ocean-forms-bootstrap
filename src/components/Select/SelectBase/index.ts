export * from './SelectBase';
export * from './SelectBase.types';
