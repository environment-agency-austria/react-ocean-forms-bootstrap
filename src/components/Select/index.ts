export * from './SelectBase';
export * from './Select';
export * from './CreatableSelect';
