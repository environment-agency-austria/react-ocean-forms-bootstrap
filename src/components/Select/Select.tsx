/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import { withField } from 'react-ocean-forms';
import { default as ReactSelect } from 'react-select';
import { Input as StrapInput } from 'reactstrap';

import { FieldLine } from '../FieldLine';
import { ISelectOption, ISelectProps, isSelectOption } from './Select.types';

/**
 * Component for displaying bootstrap
 * form groups with an select input and
 * oForm support
 */
export class BaseSelect extends React.Component<ISelectProps> {
  public static displayName: string = 'Select';

  // tslint:disable-next-line:typedef
  public static defaultProps = {
    multi: false,
    placeholder: 'ojs_select_placeholder',
  };

  /**
   * Manually handle the onChange event of
   * the select because oForms is expecting
   * default input onChange behaviour.
   */
  private handleChange = (value: ISelectOption): void => {
    const { field } = this.props;

    field.onChange({
      target: {
        value,
      },
    });
  }

  /**
   * Manually handle the onBlur event of
   * the select because oForms is expecting
   * default input onBlur behaviour.
   */
  private handleBlur = (): void => {
    const { field } = this.props;

    field.onBlur();
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const {
      field,
      placeholder,
      options,
      multi,
      meta,
    } = this.props;

    // Generate a css class based on the validity of the select element
    let selectClass = 'react-select-control';
    selectClass = (selectClass + (meta.valid ? '' : ' is-invalid')).trim();

    // Check if the current value has a different label than the value
    // with the same key in the options array. Bugfix to change the
    // selected label when the current language changes.
    let fieldValue: ISelectOption | undefined;
    if (isSelectOption(field.value)) {
      fieldValue = field.value;
      // tslint:disable-next-line:no-non-null-assertion
      const selectableValue = options.find(item => item.value === fieldValue!.value);
      if (selectableValue !== undefined && selectableValue.label !== field.value.label) {
        field.value.label = selectableValue.label;
        this.handleChange(field.value);
      }
    }

    // Support for plaintext display
    if (meta.plaintext) {
      let displayValue = '';
      if (isSelectOption(field.value)) {
        displayValue = field.value.label;
      }

      return (
        <FieldLine {...this.props}>
          <StrapInput {...field} value="" plaintext>
            {displayValue}
          </StrapInput>
        </FieldLine>
      );
    }

    const placeholderText = meta.stringFormatter(placeholder);

    return (
      <FieldLine {...this.props}>
        <ReactSelect
          id={field.id}
          value={fieldValue}
          isDisabled={field.disabled}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
          options={options}
          isMulti={multi}
          placeholder={placeholderText}
          className={selectClass}
        />
      </FieldLine>
    );
  }
}

export const Select = withField(BaseSelect);
