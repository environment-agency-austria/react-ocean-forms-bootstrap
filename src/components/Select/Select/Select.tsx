/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { default as ReactSelect } from 'react-select';

import { IPreparedSelectProps, SelectBase, ISelectFieldValue } from '../SelectBase';
import { ISelectProps } from './Select.types';

/**
 * Renders the actual `Select` component with the passed props
 */
const renderSelect = (preparedProps: IPreparedSelectProps): JSX.Element => {
  return (
    <ReactSelect {...preparedProps} />
  );
}

/**
 * Component for displaying bootstrap
 * form groups with an select input and
 * oForm support
 */
export const Select = <TSubmitValue extends unknown = ISelectFieldValue>(props: ISelectProps<TSubmitValue>): JSX.Element => {
  const {
    ...selectBaseProps
  } = props;

  return (
    <SelectBase
      {...selectBaseProps}
      renderSelect={renderSelect}
    />
  );
};
