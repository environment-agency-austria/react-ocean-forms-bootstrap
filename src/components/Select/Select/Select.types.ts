/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Subtract } from 'react-ocean-forms';

import { ISelectBasePropsBase, SelectBase } from '../SelectBase';

type ISelectBaseDefaultProps = (typeof SelectBase)['defaultProps'];
type WithExcluded = Subtract<ISelectBasePropsBase, ISelectBaseDefaultProps>;

/**
 * Props for the component `Select`
 */
export interface ISelectProps extends WithExcluded, Partial<ISelectBaseDefaultProps> {
}
