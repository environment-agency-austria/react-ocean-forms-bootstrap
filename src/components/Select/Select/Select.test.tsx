import * as React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { IFieldComponentFieldProps, IFieldComponentMeta } from 'react-ocean-forms';
import { SelectBase as ReactSelect } from 'react-select';

import { createMockField, createMockFieldMeta } from '../../../test-utils/enzymeFormContext';
import { ISelectOptions, SelectBase } from '../SelectBase';
import { BaseSelect } from './Select';
import { ISelectProps } from './Select.types';

describe('<Select />', () => {
  interface ISetupArgs {
    props?: Partial<ISelectProps>;
    fieldOverrides?: Partial<IFieldComponentFieldProps>;
    metaOverrides?: Partial<IFieldComponentMeta>;
  }

  interface ISetupResult {
    wrapper: ShallowWrapper;
    field: IFieldComponentFieldProps;
    meta: IFieldComponentMeta;
    options: ISelectOptions;
  }

  const setup = ({
    props,
    metaOverrides,
    fieldOverrides,
  }: ISetupArgs = {}): ISetupResult => {
    const fieldLabel = 'field0';
    const meta = {
      ...createMockFieldMeta(),
      ...metaOverrides,
    };
    const field = {
      ...createMockField(),
      value: '',
      ...fieldOverrides,
    };

    const options = [
      { value: 'one', label: 'One' },
      { value: 'two', label: 'Two' },
    ];

    const wrapper = shallow((
      <BaseSelect
        label={fieldLabel}
        meta={meta}
        field={field}
        options={options}
        {...props}
      />
    ));

    return {
      wrapper,
      field,
      meta,
      options,
    };
  };

  it('should render correctly', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a select-base', () => {
    const { wrapper } = setup();

    expect(wrapper.find(SelectBase).exists()).toBeTruthy();
  });

  it('should render a react-select', () => {
    const { wrapper } = setup();

    type SelectInstance = { renderSelect(): JSX.Element };
    const i = ((wrapper.instance()) as unknown) as SelectInstance;
    const result = shallow(i.renderSelect());

    expect(result.find(ReactSelect).exists()).toBeTruthy();
  });
});
