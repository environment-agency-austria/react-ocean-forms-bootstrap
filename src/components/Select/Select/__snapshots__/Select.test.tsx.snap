// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render correctly 1`] = `
<SelectBase
  field={
    Object {
      "disabled": false,
      "id": "mock-field",
      "name": "mock-field",
      "onBlur": [MockFunction],
      "onChange": [MockFunction],
      "value": "",
    }
  }
  label="field0"
  meta={
    Object {
      "error": null,
      "isRequired": false,
      "isValidating": false,
      "plaintext": false,
      "stringFormatter": [MockFunction],
      "touched": false,
      "valid": true,
    }
  }
  multi={false}
  options={
    Array [
      Object {
        "label": "One",
        "value": "one",
      },
      Object {
        "label": "Two",
        "value": "two",
      },
    ]
  }
  placeholder="ojs_select_placeholder"
  renderSelect={[Function]}
/>
`;
